### DISCUSSION TITLE
How to enumerate from a custom data feed

### DISCUSSION THREAD

**Discussion by mwpowellhtx** [OrchardCMS/OrchardCore#11368] [Q&A]
Let's say we have a _spreadsheet_ asset, or _database_, or perhaps even an external, potentially dynamic _API_, that delivers _record oriented data_ i.e. `IEnumerable<TSource>` for integration within the _CMS framework_. How might we enumerate these, doing an appropriate _ETL_, perhaps, from the source data set to the _OrchardCore/CMS framework_? Ostensibly, I am assuming this would land as at least a `ContentItem`, or derivation thereof? And however that got organized, as a _bag_, _list_, etc, as appropriate. Assuming that would require a custom module, for starters, which knowledge of that data source. However, I am not familiar with the _OrchardCore CMS_ vis-a-vis `ContentItem` hierarchy, or how to customize them, around such an objective.

**Comment by Skrypt** [comment]
You can import the spreadsheet as is in the Orchard Core database. Let's say, SpreadSheetTable. Then in the SQL Query module you can create a SQL Query that will return the data you require to use. Moreover, you can use the Schema on the SQL Query to create a GraphQL endpoint to that custom table. Of course, this is the simplest way to do it.

Though, if you want to create a ContentItem you will need to create a recipe.json file out of your SpreadSheet data. You start by creating a Content Type that will have the same column dbtypes than your spreadsheet and then you create 1 content item from the admin UI. You use the `Import/Export` feature in OC to export the data as a recipe.json file and then you base yourself on this JSON data to create your spreadsheet migration code to a OC recipe.

**Reply by mwpowellhtx** [reply]
Well, our database is _postgresql_, or at least this is the current assumption, for many reasons, not least of which we'd like to leverage `MartenDB` at some level, for our internal modules if/when appropriate to do so. Assuming the queries and such are analogous using the [`Npgsql`](https://www.nuget.org/packages/Npgsql) package.

I think, however, the latter half of the response is precisely what we want to avoid. We need/want for that to be entirely dynamic, _automated_ if you will; i.e. the data source can/should change at a moment's notice, yet that just seems still tied to the admin editors, with a manual step in between.

**Reply by Skrypt** [reply]
If you are using Content Types to define your data structure then it is dynamic. And it can change on runtime. So, there is no automation that will change your Views to add a new field or add a new content part to your views. Though, we have Admin UI templates to create Liquid Views from runtime.

Of course, when there is a custom table required then you generally will create a Migration.cs file to automate the data transformation. Then, of course, you will require some Entity class that binds to YesSQL. I don't think we have documentation about this but we have examples in the source code.

Knowing that you want to use MartenDB at some point makes this a little more difficult because right now I can't give you support about that. We would need to take a look at MartenDB and evaluate how it can be used in the context of Orchard Core. Of course, nothing prevents you from using MartenDB in a custom OC module with your own DataProvider and ... you get the idea.

**Reply by mwpowellhtx** [reply]
I imagine the core enabling technology, _postgresql_ support for `json` (`bson`) is the underpinning thing. Beyond that, don't know the particulars. I imagime some of the table scaffold around that is a bit different comparing/contrasting `YesSQL` versus `MartenDB`. I also like `MartenDB` because it can be an event source, which is one of its selling points IIRC.

I will have to explore the content types. My experience there so far is a bit weak; confidence about 70%. What happens when we try to introduce a custom type, part, etc... [The editor throws an exception](https://github.com/OrchardCMS/OrchardCore/issues/11314). Not sure what the resolution is there.

**Reply by Skrypt** [reply]
Take a look at the Workflow module source code to see some examples of custom tables usage. Of course, this uses YesSQL though. I believe that PostgreSQL can make use of JSON functions to retrieve the data directly with SQL queries. Though, Orchard Core doesn't make use of these JSON functions. We do a map/reduce design pattern with our tables. So, the Document table contains the JSON data in the Content column and our SomethingIndex tables are containing the mapped data that is required to filter on. Also, we use SomethingIndex tables for performance reasons too. I can't comment on MartenDB, I never used it.

**Reply by Skrypt** [reply]
I'm going to take some time this afternoon to read the MartenDB documentation.

**Reply by mwpowellhtx** [reply]
@Skrypt I appreciate the eyes; MartenDB aside, I need to know more about what sort of shapes we're talking about, what is the goal transformation, given say the Agency recipe? Out of the box there is the static content, right. What do we need to provide for that to be dynamic? Is it a JSON thing? A liquid thing? I'm not sure the examples are that clear on the source domain model.

**Reply by mwpowellhtx** [reply]
Considering approaches here. There may be some advantages to enumerating items dynamically during an map reduce event sequence, if possible. For instance, potentially helping with SEO and so forth.

Alternate to that, to be truly dynamic, and I mean dynamic in the run time space, if at all possible. Perhaps less helpful with the SEO, but it truly helps us by engaging visitors with content that is rendered dynamically.

Forums, blogs, etc, is one potentially application here. But also for the _'boilerplate'_ Agency recipe sections, where we might want to update or adjust services, portfolio, etc, in a regular, arbitrary manner.

**Reply by mwpowellhtx** [reply]
@Skrypt Need a bit of orientation here. Which moving parts do I need? _Enable Queries_ from the admin features? For starters? Along these lines? Thank you...

**Comment by Skrypt** [comment]
https://martendb.io/introduction.html

To be honest. I'm watching this video https://sec.ch9.ms/ch9/2d29/a281311a-76bb-4573-a2a0-2dd7affc2d29/S315dotNETconf_high.mp4 about Marten and it is really similar to YesSQL. In fact, they are using the same tech for database persistence which is Dapper. The only difference is that they are using JSON functions in PostgreSQL to fetch the data whereas YesSQL doesn't. Though YesSQL has a lot of the same things that Marten has.

I have noticed that in that video he says that they didn't implement a solution with SQL Server because of the lack of these same JSON path functions but `SQL Server` has them since `SQL Server 2016`. So maybe since the video is from 2017 they didn't know at that time.

Things that I noted that YessSQL doesn't have : 

Compiled Queries (faster than LINQ queries)
Patching API (uses JSON functions)
Events (As in DB events to get/set raw JSON documents)

@sebastienros I think Marten works a lot like YesSQL. The Patching API (json functions) is something I already talked about before. Maybe it would be worth starting to consider having it with Dialects for every DB Providers that allows it. 

Actually I think that Marten decided to support only PostgreSQL because it was easier to only support the JSON functions from that DbProvider only whereas YesSQL tries to support every DBProviders.

**Reply by mwpowellhtx** [reply]
@Skrypt Up to y'all if/when/how you'd want to pursue _Marten_. As we've mentioned, I'd probably contain my domain models to modules with that as the data tier serialization for the time being.

But for purposes of this discussion, and our pursuits with _OC/CMS_, I'm more concerned about engaging with its domain model vis-a-vis _content parts_, _items_, _types_, and so forth, and how we might go about bridging that transformation. Examples based on the _Agency_ recipe, for instance, would be helpful.

**Comment by mwpowellhtx** [comment]
I am turning a couple of choices over in my mind how to approach this. Is it worth the trouble transforming our enumerated domain model, whatever its source, to _Content Items_ (ostensibly, _presumably_, ...), or do we simply present it as a custom _Content Part_, with its view doing the enumeration. And as for its _Edit Views_, ..., well we have the same block considering that as for my [initial report](https://github.com/OrchardCMS/OrchardCore/issues/11314). So we really do need to find what the heck is going on with the out of the box custom part, edit views, etc. Maybe it is a typo, I don't know. Also, not sure we want liquid in the near/medium/long range. Probably we just need/want the Razor, for a host of reasons, but that is neither here nor there

**Comment by mwpowellhtx** [comment]
@Skrypt I am wondering, [possibly something along these lines](https://docs.orchardcore.net/en/latest/docs/reference/core/Data/#handling-prefixed-tables) for the _Part_ `HomeController`? Except _instead of a database connection_, reaching out to _Excel_, or other, supporting assets.

**Reply by Skrypt** [reply]
Executing pure SQL will work this way. Though, it will not be SQL that works with any RDB as when you are using the ContentManager that passes through YesSQL. That's the only drawback.

**Reply by mwpowellhtx** [reply]
Also not sure it was mentioned here, but elsewhere, something along the lines [what Lombiq is doing re: Person](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Controllers/PersonListController.cs#L151), perhaps. I see that controller is doing some work with _Content Manager_, which might possibly work, assuming we can connect properly with a workflow event, let's say.
